### PR TITLE
Setup retry for System.Net.* tests

### PR DESCRIPTION
**Pull request by ChadNedzlek**
The Helix SDK in arcade supports test retries in Helix by reading this file, "test-configuration.json" if it's present.

This sets up any test who's assembly name starts with System.Net.* to have 3 attempts (2 reruns) locally.

### CODE REVIEW

**Comment by ChadNedzlek** [comment]
@danmoseley Any thoughts/concerns about using this pattern for an initial go at test retries? I noticed in our telemetry these tests have a relatively high failure rate, so they seemed like a good place to start.

**Comment by ViktorHofer** [comment]
> @danmoseley Any thoughts/concerns about using this pattern for an initial go at test retries? I noticed in our telemetry these tests have a relatively high failure rate, so they seemed like a good place to start.

That's a question for @karelz as well.

> This sets up any test who's assembly name starts with System.Net.* to have 3 attempts (2 reruns) locally.

As you are mentioning the word "locally", does that imply that it works in both Helix and when running the tests locally ie with dotnet test?

**Comment by ChadNedzlek** [comment]
@ViktorHofer, sorry, no. "Locally" refers to the fact that the retries are all run on a single Helix machine (we don't attempt to coordinate retries between machines).

I'll try and write up a doc in arcade explaining how this all works.

**Comment by ChadNedzlek** [comment]
Here is a documentation PR explaining the configuration files, feedback so I can answer any questions is appreciated:
https://github.com/dotnet/arcade/pull/7749

**Comment by ChadNedzlek** [comment]
There is a potential future "non-local" retry, where we might try to run the tests on a different machine, in case the test failures are due to machine state. But that isn't implemented yet, because it's rather elaborate, and we want to make sure that it's a good use or resources, or if "local" re-execution will get our pass rates to a good place on it's own.

**Comment by danmoseley** [comment]
This seems OK to me (excited to try this) but @ManickaP in @karelz absence should sign off.

If this works, it would be nice to have it in 6.0 before branching, since I guess that will make servicing easier.

**Comment by danmoseley** [comment]
@akoeplinger there is a crash in Mono here with dumps. Test is Loader/classloader/regressions/523654/test532654_b/test532654_b.sh

**Comment by stephentoub** [comment]
> Any thoughts/concerns about using this pattern for an initial go at test retries?

My hesitancy here has been that we're casting a very wide net and could end up hiding non-deterministic failures that are actually due to product bugs (e.g. race conditions).  Will we generate reports or something for tests that are frequently retried?

**Comment by ViktorHofer** [comment]
Apart from what @stephentoub expressed, what worries me is that this introduces a difference on how we run tests on CI vs locally. How would we make sure that we don't regress the local developer test experience when enabling this feature? Is it in scope of this effort to enable test retries locally as well?
Side question, as we want to move .NET Framework testing from Helix back to the build agent (https://github.com/dotnet/runtime/issues/53924), would that feature work there as well?

**Comment by ManickaP** [comment]
@ChadNedzlek is there any record of the retried tests and they're failures for which they were retried? Kusto for instance? Just in case we wanted to analyze the data or on the contrary exclude them from internal reports.

I'm personally not oppose to this change, we should at least try it, @karelz is back though :smile:

**Comment by antonfirsov** [comment]
I agree with the concerns that in it's current form this is a dangerous proposal that may lead to false greens by hiding sporadic bugs.

It would be much better, if we could enable retrying on per-test basis, so we could selectively use it only on tests where we know that retrying can help dealing with some environmental flakyness, without risking false greens:

```C#
[RetryCount(3)]
[Fact]
public void Test() { }
```

Somewhat off-topic, but an actual game-changer in networking tests would be an ability to disable parallel execution on per-test basis, since many false-reds are rooted in timing issues caused by parallel workloads, and the current `[TestCollection]` based isolation leads to [very ugly and costly tricks](https://github.com/dotnet/runtime/tree/main/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive).

```C#
[NonParallel]
[Fact]
public void Test() { }
```

**Comment by stephentoub** [comment]
> the current [TestCollection] based isolation leads to very ugly and costly tricks

Costly?

**Comment by antonfirsov** [comment]
> Costly?

1. In terms of implementation and maintenance: splitting up a test class (or an entire hierarchy of test classes in case of sockets) just because some of the tests need to be run sequentially. The split usually counter-intuitive, since it doesn't have to do anything with the normal logic of organizing tests.
2. In terms of test execution time: if we have a test class that mostly contains tests we prefer to run sequentially, but some of them are fine to be run in parallel, and we want to avoid bothering splitting up the test class because of a few tests, we end up with longer test runs.

**Comment by ChadNedzlek** [comment]
We definitely can do it on a test by test basis, but there are hundreds of tests that would have to be constantly micromanaged in that situation, and will likely lead to a weak improvement to PR reliability, as you'd be hitting tests that just hadn't be marked yet.

Do you have a more specific set of tests in mind?  When I looked at these assemblies, there are thousands of failures in them in the last month, which is why I chose this pattern.

We are working on reporting now, so that we can produce weekly (or some other timeframe reports) around test failures rates.  Right now, a flaky tests doesn't get caught in an appropriate PR a pretty wide swath of the times, leading people to not trust the PR validation system, often merging on red, or just hitting the retry button anyway (which hides the failure in a way we won't be able to report on).

Basically this just automates what people are already doing, but it in a way that uses far less resources, reports back to your PR the real state (you can query for "Passed on Rerun" tests, but not "someone reran the build and the test stopped failing tests), and in a way we will, shortly, have detailed reporting around.

**Comment by ChadNedzlek** [comment]
Also, this is just a step in the right direction.  We can always pull the scoping back later if it turns out it's wasting time/hiding things, but without an attempt, we won't know.  It will also greatly improve the upcoming reporting effort if we have actual data to report on, rather than just blind guesses that might lead to a really unpleasant situation.

Can we agree to turn this on for a month or two, and we can look back at the data to see what's come of it (and how it looks with the reporting information)?

I don't know that we can go from here to perfect in one step, and I'm hoping this one is at least one in the right direction (and one we can always step-back later if we need to)

**Comment by bartonjs** [comment]
> Can we agree to turn this on for a month or two, and we can look back at the data to see what's come of it 

FWIW: So long as data is being gathered, I'm supportive.  In an ideal world I'd love to see all flaky tests (even my own) get "un-flaked", but I think being in a place where people actually look at errors before blindly hitting retry is better than where we are now.

Though I'd wait until after the fork... start 7 off with new science, but don't let things hide too much in 6-RC.

**Comment by karelz** [comment]
This is great, and I am looking forward to have retries.
I would like to clarify 2 things first:
1. If a test fails, but passes on retry, how will it look in the "checks" of PR? Is there a screenshot? Will that indicate that it passed only on retry, or does one have to go into AzDO details to find out?
2. Will the first-run data be reported to Kusto the same way as today? ... We are actively monitoring Kusto to spot regressions, outages and flaky tests across runs, incl. PRs. We need to have all failures recorded there, otherwise we will run blind and will not know when our tests start regressing - pushing the problem just a bit down the road.

On a side note, I would say that 1 retry should be enough. I would start with that ...

**Comment by ChadNedzlek** [comment]
Answers for @karelz 
1. In the checks tab, right now, they won't show up at all in passed on retry state.  It's a lot of work to fetch those results, and we weren't sure about the benefit of showing them there, and how it would be clear which tests are blocking and which are not, so there's a lot of UI work to consider there too.  If that reporting is important, we can work at that.  Do you consider that a blocker for this work?
2. At the moment, they won't show up in Kusto at all (because they will no longer be failing).  We are working on aggregate reporting work for our next push, so that we can show more detailed information about these flaky tests (and non-flaky tests), but we need something to actually use the feature so that we have data to report on.  If nothing is using retry, we won't have any data with which to build/test the reports, so it's likely we will create something unusually bad.  I was hoping we could use System.Net to get some data flowing through the system so we could get some real-world data with which to help drive our designs, and also help PR pass rates up a bit in the process.

Is it ok to get this in now, with the expectation that reporting will shortly follow, and knowing that we can revert it if something goes badly?  We should be early enough in the 7 frame that I was hoping we'd have the flexibility to experiment a bit in this space with time to change course if something isn't working out.

**Comment by ViktorHofer** [comment]
@ChadNedzlek @markwilkie do we have an answer to what I asked previously in this thread?

> Apart from what @stephentoub expressed, what worries me is that this introduces a difference on how we run tests on CI vs locally. How would we make sure that we don't regress the local developer test experience when enabling this feature? Is it in scope of this effort to enable test retries locally as well?
Side question, as we want to move .NET Framework testing from Helix back to the build agent (#53924), would that feature work there as well?

Making this a helix only feature will prevent us moving workloads like .NET Framework testing back to the build agent for which we don't need Helix at all. Also making this Helix only means that we don't have a metric on how stable our tests are when being run locally without the retry feature being present. Are these concerns known and are there any plans to mitigate them?

**Comment by markwilkie** [comment]
@ViktorHofer and I chatted and his point that ideally we'd be at a place where we could run our tests in a common/shared fashion - but without relying on "helix" per-say is a good one.  (Helix in this case is the infra which runs arbitrary work loads in the cloud)  We also agreed that we should make sure this is tracked, but also should not block this particular PR.

This work for shared test infra is being tracked by https://github.com/dotnet/arcade/issues/6948 and is under consideration for .NET 7.  (although it'll be quite expensive for the product teams)

@karelz - I see @ChadNedzlek responded to your questions, thoughts on going forward with this PR?  Basically we're looking for a narrow/targeting set of tests to turn on the feature such that we can gather data and report such that we don't hide real issues.

**Comment by karelz** [comment]
@markwilkie I have concerns about the answers to question [2] above - that may have significant impact on us. We will meet with @ChadNedzlek tomorrow to discuss our CI monitoring workflow. We will be able to say more after that.

**Comment by karelz** [comment]
We had more discussion on Networking team, and discussion with @ChadNedzlek -- my key takeaway is that **until RC2 snaps**, not having Kusto reporting on 1st failures is **blocker for us**. After RC2 we'd be fine with limited time period (e.g. 1 month) to be in such state (not having Kusto repots on 1st failures).

This upcoming month is sensitive, because if we make a fix we want to port to 6.0 and the fix introduces more test/product flakiness (e.g. 1 failure in 10 runs, which is pretty bad), Kusto is the only place we will learn about it prior to porting to 6.0 branch failures (which would be bad).

**Comment by ChadNedzlek** [comment]
That makes perfect sense to me, and I'm working on the code to pull the information into Kusto right now, so it should be available in a couple weeks in prod.

**Comment by markwilkie** [comment]
Thanks @karelz and @ChadNedzlek - this is good thinking.

**Comment by karelz** [comment]
Great, thanks @ChadNedzlek and @markwilkie. In the meantime, I will close this PR (to avoid sitting on stale PRs ;))
We can reopen when either Kusto has 1st failure reporting, or we are good for RC2.

**Comment by markwilkie** [comment]
@ChadNedzlek write up after talking with @karelz.

Chad NedzlekSummary of meeting with Karel (in charge of networking stack)
    * particular error message patterns to "ignore" because they are not test specific
* count of failures overall, did it it go way up/down?
* a particular test failure in official/PR, did it go up/down?
* tracking bug example 57626
* two years of aggregate data can be useful, to see if something changed in frequency, can we support that? Probably for aggregations.
* hangs are complicated, because they aren't exposed out as test failures, can we do something here?
* if one PR has LOTS of failures, it's probably ignorable, because it's likely caused that that PR or an outage, neither of which is interesting for historical trends
* outages can cause "noise" too

Also, RC2 is going to be taking a lot of backports from main for the next month or two, and it's important that networking quality stay high, so it's not the ideal time to experiment.  We either need the data in Kusto, or to wait for RC2 to stabilize before we move forward.  I think we can probably get the data in Kusto pretty quickly, since it doesn't have a bunch if UI to mess with.

**Comment by ChadNedzlek** [comment]
We should be storing all the data that blocked our previous attempt at this PR.  I'm going to reopen this so we can consider it now.

**Review by karelz — APPROVED** [review]

**Comment by ChadNedzlek** [comment]
Quick update.  Since this was checked it, it's trigged in 28 builds. 9 of those builds were changed to complete successes by retrying these assemblies. It seems to be working, and we'll try and look at some data to see if there are other assemblies/tests that we might consider.

**Comment by ManickaP** [comment]
@ChadNedzlek how are the data stored in Kusto? I'm seeing some weird duplicities/result when I join TestResults with Jobs and/or WorkItems. Could this change be a root cause of this?

**Comment by markwilkie** [comment]
You can try https://github.com/dotnet/core-eng/blob/main/Documentation/Project-Docs/Dev%20Workflow/Passed%20On%20Rerun%20Data.md @ManickaP.  I'm sure @ChadNedzlek  can help too if the docs don't.  

Also - if the docs are insufficient, it be great to get feedback so we can improve them.

**Comment by antonfirsov** [comment]
Is it possible to disable reruns in individual `/azp run ...` commands which we do on PR-s? It would be valuable to catch sporadic test failures when PR-ing networking code changes. (Otherwise it is easier to merge flaky tests or even bugs.)

**Comment by markwilkie** [comment]
I don't think a disable-per-run is supported yet.  (obviously the .json could updated....but I know that's a pain)

It's a good suggestion though, would you mind adding an issue for us?

**Comment by ChadNedzlek** [comment]
@ManickaP The table that you'll want to look at is AzureDevOpsTests (TestResults is an old mechanism that we'll remove eventually).  However, duplicates in Kusto are expected, because of the nature of our services, so you'll often want to summarize/distinct to handle them.

@antonfirsov  If you are concerned with reruns, you can always check AzDO for them.  The default filter for tests is Failed + Aborted, but you can always change that filter to include "Passed on rerun", and look at your options.

The filter button isn't always easy to find, here's a picture of it:
image

**Comment by antonfirsov** [comment]
@ChadNedzlek thanks! That filter is addressing my concern and seems to be easy enough to work with.

**Comment by ManickaP** [comment]
When I try to join AzureDevOpsTests table with WorkItems I get: `Inconsistent data types for the join keys (WorkItemId, WorkItemId) (I32, I64)` But that's probable for a separate issue in core-eng.

**Comment by ManickaP** [comment]
Thanks @ChadNedzlek!
